### PR TITLE
Added solution for Project Euler problem 144

### PR DESCRIPTION
**Pull request by fpringle**
### Problem 144: [Investigating multiple reflections of a laser beam](https://projecteuler.net/problem=144)

In laser physics, a "white cell" is a mirror system that acts as a delay line for the laser beam. The beam enters the cell, bounces around on the mirrors, and eventually works its way back out.

The specific white cell we will be considering is an ellipse with the equation 4x<sup>2</sup> + y<sup>2</sup> = 100

The section corresponding to −0.01 ≤ x ≤ +0.01 at the top is missing, allowing the light to enter and exit through the hole.

The light beam in this problem starts at the point (0.0,10.1) just outside the white cell, and the beam first impacts the mirror at (1.4,-9.6).

Each time the laser beam hits the surface of the ellipse, it follows the usual law of reflection "angle of incidence equals angle of reflection." That is, both the incident and reflected beams make the same angle with the normal line at the point of incidence.

In the figure on the left, the red line shows the first two points of contact between the laser beam and the wall of the white cell; the blue line shows the line tangent to the ellipse at the point of incidence of the first bounce.

The slope m of the tangent line at any point (x,y) of the given ellipse is: m = −4x/y

The normal line is perpendicular to this tangent line at the point of incidence.

The animation on the right shows the first 10 reflections of the beam.

How many times does the beam hit the internal surface of the white cell before exiting?

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by mateuszz0000 — CHANGES_REQUESTED** [review]
Please resolve conflicts.

File: project_euler/problem_144/sol1.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    )
+    outgoing_gradient = (s2 - c2 * incoming_gradient) / (c2 + s2 * incoming_gradient)
+
+    # to find the next point, solve the simultaeneous equations:
+    # y^2 + 4x^2 = 100
+    # y - b = m * (x - a)
+    # ==> A x^2 + B x + C = 0
+    A = outgoing_gradient * outgoing_gradient + 4
+    B = 2 * outgoing_gradient * (point_y - outgoing_gradient * point_x)
+    C = (point_y - outgoing_gradient * point_x) ** 2 - 100
```
Comment (dhruvmanila) [inline project_euler/problem_144/sol1.py:63]: I don't believe these are constants, so if you could provide better names that would be awesome :)

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
Small nitpick and then this will be good to go. Thanks for doing this :)

**Comment by XmenchoX** [comment]
just stop spaming

**Review by dhruvmanila — APPROVED** [review]

**Pull request merged**